graph_altimeter/scan/iam/__init__.py

```diff
@@ -0,0 +1,309 @@
+            permissions = rule.matches(arn)
+            if permissions is None:
+                continue
+            edge = _create_edge(
+                    from_id,
+                    v["~id"],
+                    "iam_resource_link",
+                    {"permission": permissions}
+                )
+            edges.append(edge)
+    return edges
+
+
+def _create_edge(from_id, to_id, label, properties):
+    """Creates and returns and Edge between the specified vertex ids, with
+    the given labels and properties."""
+
+    # TODO: By now the neptune_client does write arbitrary properties in edges
+    # so the properties specified here are not actually persisted.
+    e = {
+        "~id": uuid.uuid1(),
+        "~from": from_id,
+        "~to": to_id,
+        "~label": label,
+    }
+    for name in properties:
+        e[name] = properties[name]
+    return e
+
+
+def _aws_arn_rules(document):
+    """Given a string containing a policy document returns the corresponding
+    list of rules."""
+    document = json.loads(document)
+    policy = Policy(document)
+    rules = []
+    for statement in policy.statements:
+        for resource in statement.resources:
+            try:
+                rule = ARNRule(
+                        statement.action_summary(),
+                        statement.actions_expanded,
+                        resource,
+                        statement.effect,
+                )
+                rules.append(rule)
+            except InvalidArnPattern as e:
+                # There could be policy documents that contain invalid
+                # policies.
+                logger.warning(e)
+    return rules
+
+
+class ARNRule:
+    """Represent a rule parsed from a policy document statement."""
+    PERMISSIONS_TRANSLATION = {
+            "Tagging": "Write",
+            "Permissions": "Write",
+            "Read": "Read",
+            "List": "Read",
+            "Write": "Write",
+    }
+
+    def __init__(self, service_permissions, actions, arn_pattern, effect):
+        """Creates a rule from a given set of service permissions, actions, a
+        arn pattern and an effect. """
+        # Service permissions example:
+        # {'sqs': {'Tagging', 'Permissions',
+        # 'Read', 'List', 'Write'}, 'ssm': {'Write'}}
+        # actions example:
+        # {'sqs:tagqueue','dynamodb:getitem'}
+        # arn pattern example:
+        # arn:aws:ssm:eu-west-1:*:parameter/grafana/loki/*
+
+        # Translate the permissions to Read and Write.
+        translated = {}
+        for tech in service_permissions:
+            permissions = {
+                ARNRule.PERMISSIONS_TRANSLATION[permission] for
+                permission in service_permissions[tech]
+                if permission is not None
+            }
+            translated[tech] = permissions
+            pattern = ARN(arn_pattern)
+
+        if pattern.error and pattern.arn != "*":
+            raise InvalidArnPattern(arn_pattern)
```

Comment:
Can you add a comment explaining why `pattern.error` is ignored for `*`?

Reply:
okey